homeassistant/components/tedee/coordinator.py

```diff
@@ -109,3 +102,32 @@ async def _async_update(self, update_fn: Callable[[], Awaitable[None]]) -> None:
             raise UpdateFailed("Error while updating data: %s" % str(ex)) from ex
         except (TedeeClientException, TimeoutError) as ex:
             raise UpdateFailed("Querying API failed. Error: %s" % str(ex)) from ex
+
+    def _async_add_remove_locks(self) -> None:
+        """Add new locks, remove non-existing locks."""
+        if not self._locks_last_update:
+            self._locks_last_update = set(self.tedee_client.locks_dict)
+
+        if (
+            current_locks := set(self.tedee_client.locks_dict)
+        ) == self._locks_last_update:
+            return
+
+        # remove old locks
+        if removed_locks := self._locks_last_update - current_locks:
+            _LOGGER.debug("Removed locks: %s", ", ".join(map(str, removed_locks)))
+            device_registry = dr.async_get(self.hass)
+            for lock_id in removed_locks:
+                if device := device_registry.async_get_device(
+                    identifiers={(DOMAIN, str(lock_id))}
+                ):
+                    device_registry.async_remove_device(device.id)
```

Comment:
It's generally safer to just remove the config entry id from the device, via `device_registry.async_update_device`. If there are no other config entries connected to the device it will then be removed by the device registry.

https://github.com/home-assistant/core/blob/d0e6ce193cd7729b23fd6d4ee229a997f4548699/homeassistant/helpers/device_registry.py#L644